connectors/sources/jira.py

```diff
@@ -0,0 +1,467 @@
+                "value": 50,
+                "label": "Number of concurrent downloads for fetching attachment content",
+                "type": "int",
+            },
+        }
+
+    def tweak_bulk_options(self, options):
+        """Tweak bulk options as per concurrent downloads support by jira
+
+        Args:
+            options (dictionary): Config bulker options
+
+        Raises:
+            Exception: Invalid configured concurrent_downloads
+        """
+        if self.concurrent_downloads > MAX_CONCURRENT_DOWNLOADS:
+            raise Exception(
+                f"Configured concurrent downloads can't be set more than {MAX_CONCURRENT_DOWNLOADS}."
+            )
+        options["concurrent_downloads"] = self.concurrent_downloads
+
+    def _validate_configuration(self):
+        """Validates whether user input is empty or not for configuration fields
+
+        Raises:
+            Exception: Configured keys can't be empty
+        """
+        logger.info("Validating Jira Configuration...")
+        connection_fields = (
+            ["host_url", "service_account_id", "api_token"]
+            if self.is_cloud
+            else ["host_url", "username", "password"]
+        )
+
+        if empty_connection_fields := [
+            field for field in connection_fields if self.configuration[field] == ""
+        ]:
+            raise Exception(
+                f"Configured keys: {empty_connection_fields} can't be empty."
+            )
+
+        if not self.ssl_disabled and self.certificate == "":
+            raise Exception("When ssl_disabled is False then ssl_ca can't be empty.")
+
+    def _generate_session(self):
+        """Generates a new session"""
+        logger.debug("Creating a session...")
+        if self.is_cloud:
+            auth = (
+                self.configuration["service_account_id"],
+                self.configuration["api_token"],
+            )
+        else:
+            auth = self.configuration["username"], self.configuration["password"]
+
+        basic_auth = aiohttp.BasicAuth(
+            login=auth[0],
+            password=auth[1],
+        )
+        request_headers = {
+            "accept": "application/json",
+            "content-type": "application/json",
+        }
+        timeout = aiohttp.ClientTimeout(total=None)
+        return aiohttp.ClientSession(
+            auth=basic_auth,
+            headers=request_headers,
+            timeout=timeout,
+            raise_for_status=True,
+        )
+
+    async def close(self):
+        """Closes unclosed client session"""
+        if self.session is None:
+            return
+        await self.session.close()
+
+    async def _api_call(self, url_name, is_pagination=False, **url_kwargs):
+        """Method for adding retries whenever exception raised during an api calls
+
+        Args:
+            url_name (str): Jira query name to be executed.
+            is_pagination (bool): Flag to check if the pagination is required
```

Comment:
Why can it be not required?

Reply:
Since, Atlassian APIs does not provide pagination for fetching projects and returns all projects at a time. Also, pagination is not needed for ping since it's using `\myself` endpoint. So, it is only required for fetching the Jira issues.

Reply:
In this case I advise to separate this method into two - one with pagination, one without. They can both use same private methods to simplify logic.

Flags that control the flow of method that drastically is normally an anti-pattern

Reply:
I see, thanks for the suggestion. I'll update the code accordingly.